keras/src/saving/saving_lib.py

```diff
@@ -202,6 +227,8 @@ def _load_model_from_fileobj(fileobj, custom_objects, compile, safe_mode):
         weights_store.close()
         if asset_store:
             asset_store.close()
+        if weights_file_path:
```

Comment:
We should put the `unlink` in a try/except/finally block to make sure it gets executed no matter what.

Reply:
I have added a try/finally block.
In `_save_model_to_fileobj`, if an error is raised, the bad `model.weights.h5` won't be written to the zip file.
In `_load_model_from_fileobj`, if an error is raised, the extracted `model.weights.h5` will be deleted no matter what.

Tests have been added to verify these behaviors.